### ISSUE TITLE
add onlyone plugin

### ISSUE THREAD

**Issue by fabrizior** [coredns/coredns#4077]
I'm not sure this is the right place where ask, so excuse me if I'm wrong. I'm trying to work with CoreDNS plugins, I would like to add the onlyone plugin, but following the instruction from "Learning CoreDNS", of course, I've made some mistakes.
I inserted `onlyone:github.com/coredns/learning-coredns/plugins/onlyone` into the plugin.cfg file, but when I run make I have errors like _"Makefile:16: recipe for target 'CoreDNS' failed"_  
and the _go.mod_ has been changed.
Can I have some help or give me someplace where I can ask?

CoreDNS-1.7.0
linux/amd64, go1.14.6

Corefile:
```
rex.loc:5300 {
  file ./zones/db.rex.loc
}

.:5300 {
  any
  log
  forward . 8.8.8.8
}
```

@HP-ProBook:~/coredns$ cat /etc/os-release
```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Comment by fabrizior** [comment]
a bad place for the post

**Comment by miekg** [comment]
What does a "go generate" and "go build" output

**Comment by fabrizior** [comment]
This is the result.

 fabrizio@HP-ProBook:~/coredns$ go generate

go: finding module for package github.com/coredns/learning-coredns/plugins/onlyone

go: found github.com/coredns/learning-coredns/plugins/onlyone in github.com/coredns/learning-coredns/plugins/onlyone v0.0.0-20190919194013-b12b5750ab68

go: finding module for package k8s.io/api/admissionregistration/v1alpha1

fabrizio@HP-ProBook:~/coredns$ go build

go: finding module for package k8s.io/api/admissionregistration/v1alpha1

../go/pkg/mod/k8s.io/client-go@v10.0.0+incompatible/kubernetes/scheme/register.go:22:2: module k8s.io/api@latest found (v0.18.8), but does not contain package k8s.io/api/admissionregistration/v1alpha1

It looks like version incompatibility. In the original go.mod I have 
k8s.io/client-go v0.18.6.
k8s.io/klog v1.0.0

After the build go.mod is changed with  
k8s.io/client-go v10.0.0+incompatible
k8s.io/klog v1.0.0
k8s.io/utils v0.0.0-20200324210504-a9aa75ae1b89 // indirect

fabrizio@HP-ProBook:~/coredns$

 

Fabrizio
Parma

 

 

Da: Miek Gieben 
Inviato: lunedì 17 agosto 2020 18:00
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

What does a "go generate" and "go build" output

**Comment by miekg** [comment]
gah, kubernetes bs... you can just remove the kubernetes plugin, what happens then?

**Comment by fabrizior** [comment]
It looks better, but I still have one error:

 

fabrizio@HP-ProBook:~/coredns$ go generate

go: finding module for package github.com/coredns/learning-coredns/plugins/onlyone

go: found github.com/coredns/learning-coredns/plugins/onlyone in github.com/coredns/learning-coredns/plugins/onlyone v0.0.0-20190919194013-b12b5750ab68

fabrizio@HP-ProBook:~/coredns$

fabrizio@HP-ProBook:~/coredns$ go build

# github.com/coredns/learning-coredns/plugins/onlyone

../go/pkg/mod/github.com/coredns/learning-coredns/plugins/onlyone@v0.0.0-20190919194013-b12b5750ab68/setup.go:33:21: cannot use c (type *"github.com/mholt/caddy".Controller) as type *"github.com/caddyserver/caddy".Controller in argument to dnsserver.GetConfig.

 

And again, the go.mod has been modified:

fabrizio@HP-ProBook:~/coredns$ diff go.mod go.mod.ori

15d14

<       github.com/coredns/learning-coredns/plugins/onlyone v0.0.0-20190919194013-b12b5750ab68

43c42

<       k8s.io/client-go v10.0.0+incompatible

---

>       k8s.io/client-go v0.18.6

45d43

<       k8s.io/utils v0.0.0-20200324210504-a9aa75ae1b89 // indirect

 

Restori 

Parma

 

 

Da: Miek Gieben 
Inviato: lunedì 17 agosto 2020 21:30
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

gah, kubernetes bs... you can just remove the kubernetes plugin, what happens then?

**Comment by miekg** [comment]
onlyone uses github.com/mholt/caddy".Controller as import path which is wrong it should be github.com/caddyserver/caddy".Controller

**Comment by chrisohaver** [comment]
Per the project go.mod, onlyone requires coredns v1.5.0.  If you compile it into coredns 1.5.0, it _should_ work ok.

**Comment by fabrizior** [comment]
github.com/mholt/caddy v1.0.0

comes from go.mod into https://github.com/coredns/learning-coredns/plugins/onlyone/go.mod

 

I've made a local copy of onlyone plugin, but I'm not able to say how to use it. I have the book "Learning CoreDNS" where is the instructions on how to make it. It says ... modify go.mod with replace github.com/coredns/learning-coredns/plugins/onlyone => /mylocal/path-copy

 

I've changed plugin.cfg with  

onlyone:/home/fabrizio/coredns/learning-coredns/plugins/onlyone

###onlyone:github.com/coredns/learning-coredns/plugins/onlyone

 

but seems a bad syntax. Sorry I'm a beginner with go (and coredns also).

 

fabrizio@HP-ProBook:~/coredns$ go build

core/plugin/zplugin.go:7:2: package /home/fabrizio/coredns/learning-coredns/plugins/onlyone is not in GOROOT (/usr/local/go/src/home/fabrizio/coredns/learning-coredns/plugins/onlyone)

fabrizio@HP-ProBook:~/coredns$ core/plugin/zplugin.go:7:2: package /home/fabrizio/coredns/learning-coredns/plugins/onlyone is not in GOROOT (/usr/local/go/src/home/fabrizio/coredns/learning-coredns/plugins/onlyone)

-bash: syntax error near unexpected token `('

 

I'll look for: syntax error near unexpected token `('

 

 

Fabrizio

Parma

 

 

Da: Miek Gieben 
Inviato: lunedì 17 agosto 2020 21:46
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

onlyone uses github.com/mholt/caddy".Controller as import path which is wrong it should be github.com/caddyserver/caddy".Controller

**Comment by johnbelamaric** [comment]
>
>
>
> I've changed plugin.cfg with
>
> onlyone:/home/fabrizio/coredns/learning-coredns/plugins/onlyone
>
> ###onlyone:github.com/coredns/learning-coredns/plugins/onlyone
>
>
You should use the line you have commented out, and delete the line with
/home/... in plugin.cfg. The "replace" tells go to use the /home/...
directory for the named module.

>

**Comment by fabrizior** [comment]
>>>> Per the project go.mod, onlyone requires coredns v1.5.0. If you compile it into coredns 1.5.0, it should work ok.

Where. The go.mod on onlyone folder is still with coredns 1.5.0:

 

module github.com/coredns/learning-coredns/plugins/onlyone

go 1.12

 

require (

        github.com/coredns/coredns v1.5.0

        github.com/caddyserver/caddy v1.0.0

        github.com/miekg/dns v1.1.12

        github.com/prometheus/client_golang v0.9.3

)

 

I can modify it on my local folder but now I have:

fabrizio@HP-ProBook:~/coredns$ go build

core/plugin/zplugin.go:7:2: package /home/fabrizio/coredns/learning-coredns/plugins/onlyone is not in GOROOT (/usr/local/go/src/home/fabrizio/coredns/learning-coredns/plugins/onlyone)

 

:-(

 

Fabrizio

 

Da: chrisohaver 
Inviato: lunedì 17 agosto 2020 21:52
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

Per the project go.mod, onlyone requires coredns v1.5.0. If you compile it into coredns 1.5.0, it should work ok.

**Comment by chrisohaver** [comment]
> >>> Per the project go.mod, onlyone requires coredns v1.5.0. If you compile it into coredns 1.5.0, it should work ok.

Actually - i take that back - i think the caddy project name change may have been backwards-breaking (i.e. the old repo name wasn't left in place - so even versioned references broke).

**Comment by fabrizior** [comment]
Ok, this is the plugin.cfg (onlyone reference):

##onlyone:/home/fabrizio/coredns/learning-coredns/plugins/onlyone

onlyone:github.com/coredns/learning-coredns/plugins/onlyone

 

 

And this the go.mod:

module github.com/coredns/coredns

 

replace github.com/coredns/learning-coredns/plugin/onlyone => /home/fabrizio/coredns/learning-coredns/plugins/onlyone

 

go 1.13

 

require (

        github.com/Azure/azure-sdk-for-go v40.6.0+incompatible

....

....

 

but it still references to some Github (github.com/mholt/caddy):

fabrizio@HP-ProBook:~/coredns$ go generate

fabrizio@HP-ProBook:~/coredns$ go build

# github.com/coredns/learning-coredns/plugins/onlyone

../go/pkg/mod/github.com/coredns/learning-coredns/plugins/onlyone@v0.0.0-20190919194013-b12b5750ab68/setup.go:33:21: cannot use c (type *"github.com/mholt/caddy".Controller) as type *"github.com/caddyserver/caddy".Controller in argument to dnsserver.GetConfig

fabrizio@HP-ProBook:~/coredns$

 

Fabrizio

 

 

Da: John Belamaric 
Inviato: lunedì 17 agosto 2020 22:23
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

>
>
>
> I've changed plugin.cfg with
>
> onlyone:/home/fabrizio/coredns/learning-coredns/plugins/onlyone
>
> ###onlyone:github.com/coredns/learning-coredns/plugins/onlyone
>
>
You should use the line you have commented out, and delete the line with
/home/... in plugin.cfg. The "replace" tells go to use the /home/...
directory for the named module.

**Comment by fabrizior** [comment]
There is a note on notes/ coredns-1.5.2.md

The CoreDNS team has released

[CoreDNS-1.5.2](https://github.com/coredns/coredns/releases/tag/v1.5.2).

 

Small bugfixes and a change to Caddy's import path (mholt/caddy -> caddyserver/caddy). Doing

a release helps plugins deal with the change better.

 

I not very sure about its meaning

 

Fabrizio 

 

 

Da: chrisohaver 
Inviato: lunedì 17 agosto 2020 22:38
A: coredns/coredns <coredns@noreply.github.com>
Cc: fabrizior <rfc@rfc.it>; State change <state_change@noreply.github.com>
Oggetto: Re: [coredns/coredns] add onlyone plugin (#4077)

 

Per the project go.mod, onlyone requires coredns v1.5.0. If you compile it into coredns 1.5.0, it should work ok.

Actually - i take that back - i think the caddy project name change may have been backwards-breaking (i.e. the old repo name wasn't left in place - so even versioned references broke).

**Comment by chrisohaver** [comment]
> I not very sure about its meaning

It means that the caddy project changed the name of their repo after coredns 1.5.0 was released. The caddy name change forced all projects using it at the time to update their import paths. 

You can see the changes that were required to update the caddy import paths in CoreDNS here ... https://github.com/coredns/coredns/pull/2961.  After this change was merged, we released CoreDNS 1.5.2.

The same types of changes need to be made in the onlyonce repo.

**Comment by johnbelamaric** [comment]
Yeah, I originally locked it to v1.5.0 to avoid hassles from changes to
CoreDNS but repo renames are something that go just doesn't handle that
well. Try updating the dependency to CoreDNS v1.5.2 and changing the caddy
path in the go.mod and see if that works for you. If so you can submit a PR
:)
